### PR TITLE
Fixed #31867 -- Made TabularInline handling of hidden fields with view-only permissions consistent with StackedInline. 

### PR DESCRIPTION
**Pull request by antoinehumbert**
ticket-31867

…in admin when user has view-only permissions

### CODE REVIEW

**Comment by antoinehumbert** [comment]
Just saw the failing tests. I will correct the PR

File: django/contrib/admin/templates/admin/edit_inline/tabular.html [hunk 1]
```diff
@@ -50,7 +50,7 @@ <h2>{{ inline_admin_formset.opts.verbose_name_plural|capfirst }}</h2>
         {% for fieldset in inline_admin_form %}
           {% for line in fieldset %}
             {% for field in line %}
-              {% if field.is_readonly or not field.field.is_hidden %}
+              {% if not field.field.is_hidden %}
```
[the remark this answers is not in the document]
Comment (antoinehumbert) [reply django/contrib/admin/templates/admin/edit_inline/tabular.html:53]: The thought about whether or not a read_only field with a HiddenInput should be displayed is precisely the reason why I talked about "inconsistency" in my bug report. Because it's currently displayed in tabular inlines but not in stacked inlines (in the case of field being the single field of a line).

In my opinion, a field with a HiddenInput should not be displayed. So that's the reason for my change in tabular.html : the test looks more logical to me than the previous one (which I think was there just because it's the negation of the related test in the "original" <td> containing the hidden input for fields. In my original patch in the ticket, I talked about rendering an empty string for the read-only field when it is hidden, because I worked on version 2.1 (and the field dictionary representation was rendered in this <td>) ; but in the master branch, the new test introduced for the "original"<td> coped with this.

The test for the "original" <td> wants to render hidden input form fields (so adding _if not field.is_read_only_ is ok because a read_only field will not render a input form field).
The test for the "field-position" <td> is just for rendering the field value, whether or not it's a **real** form field or just the raw value (for read-only field), so it looks more logical to have _if not field.field.is_hidden_.

The PR I provided will also correct the behaviour for stacked inlines -> I will add a unittest for this.
However, there will be just a little inconsistency remaining:
* in stacked inlines, the read_only + hidden values are rendered in the html (but not visible because in a "hidden" div)
* in tabular inlines, the read_only + hidden values won't appear anyhere in the html.

I think we an leave with this inconsitency, but it would also be possible to adapt to have the same behaviour by
* rendering the <th> with "hidden" class for the column title of a hidden_field
* rendering the <td> with "hidden" class for the field cell
In such a case, may be it would make more sense to render the hidden form inputs in the related hidden cell rather than in the "original" <td>
Let me know if you think this would be better. (My first thought was to avoid as most as possible the modifications in the tabular.html template, as it would require projects redefining the template - such as django-nested-admin - to do the same modifications. But this is surely not a good reason not to do this modification if it mkes things clearer)

[the remark this answers is not in the document]
Comment (carltongibson) [reply django/contrib/admin/templates/admin/edit_inline/tabular.html:53]: Hi @antoinehumbert — thanks for the quick reply. 

If we can resolve the inconsistency, then super! But we need to be clear on the behaviour (and the behaviour change) both in the tests, and possibly the release notes. (If it's particularly obscure then we maybe don't need a release note, but if it's a behaviour someone might be **using** then a small _Breaking change_ note is appropriate.) 

We also need to make sure we didn't introduce a regression ref a9f5652113f0721a7066e359ae28d14692ea3c47. (We added a test for similar in 0aeb802cf054cb369646c871b53c93a83c1fa58a the other day.)

[the remark this answers is not in the document]
Comment (antoinehumbert) [reply django/contrib/admin/templates/admin/edit_inline/tabular.html:53]: Just pushed new unit tests for stacked inlines. (may be you can have a look at my "todo" in the added tests -> I do not know why assertInHTML does not work in thoses cases, despite the fact that searched HTML is effectively present !)

For the regression, do you have any suggestion for building such a test case ? As far as I can see, I cannot see how VariableDoesNotExist warning could be raised as the only accessed variables for a read-only field are **field.field.name** and **field.contents**, which are both defined on AdminReadonlyField.

[the remark this answers is not in the document]
Comment (antoinehumbert) [reply django/contrib/admin/templates/admin/edit_inline/tabular.html]: Hi @carltongibson , I finally changed the tabular.html template so that all fields values are rendered in HMTL, even for hidden fields. But, in the case of hidden fields, the corresponding \<th\> (containing field label) and \<td\> (containing field value) have the _hidden_ class, so that the column for hidden field is not visible by user.

This is more consistent with the behaviour of stacked inlines, which for hidden fields are always rendered in HTML, but not visible because contained in some div with the _hidden_ class.

... I also corrected the unittests for hidden fields in stacked inlines.

[the remark this answers is not in the document]
Comment (antoinehumbert) [reply django/contrib/admin/templates/admin/edit_inline/tabular.html]: I think i ll definitely add a unittest for the colspan case of non-field errors. My asumption on the fact that it should be equal to the total number of columns is incorrect. It must not take into account the hidden columns, so the new behaviour will give the same incorrect html as the former behaviour.

[the remark this answers is not in the document]
Comment (antoinehumbert) [reply django/contrib/admin/templates/admin/edit_inline/tabular.html]: Just pushed correction for correct computation of colspan for non-field error row in tabular inlines

**Comment by antoinehumbert** [comment]
Hi @carltongibson , will you have some time looking at the last modifications of this pull request ?

**Comment by carltongibson** [comment]
Hi @antoinehumbert — thanks for your patience. You're on my list! 🙂

File: tests/admin_inlines/tests.py [hunk 2]
```diff
@@ -228,6 +229,113 @@ def test_inline_hidden_field_no_column(self):
             response.rendered_content,
         )
 
+    def test_tabular_inline_hidden_field_with_view_only_permissions(self):
+        """#31867 -- Make sure content of hidden field is not visible in tabular inline when user has view-only
+        permission"""
```
Comment (jacobtylerwalls) [inline tests/admin_inlines/tests.py:234]: Wrap docstrings at 79 chars and end comments with periods.

File: tests/admin_inlines/tests.py [hunk 3]
```diff
@@ -228,6 +229,113 @@ def test_inline_hidden_field_no_column(self):
...
+        child2_ct = ContentType.objects.get_for_model(Child2)
+        permission = Permission.objects.get(codename='view_parent2', content_type=parent2_ct)
+        user.user_permissions.add(permission)
+        permission = Permission.objects.get(codename='view_child2', content_type=child2_ct)
+        user.user_permissions.add(permission)
+        self.client.login(username="user", password="pwd")
+        response = self.client.get(reverse('admin:admin_inlines_parent2_change', args=(parent.pk,)))
+        # Following test ensure that the whole line containing name + position fields is not hidden
+        self.assertContains(response, '<div class="form-row field-name field-position">')
+        # Following test ensure that the div containing the position field is hidden
```
Comment (jacobtylerwalls) [inline tests/admin_inlines/tests.py:268]: Django style is to avoid "ensure that" and similar boilerplate language; just say "The whole line containing ... is not..."

File: tests/admin_inlines/tests.py [hunk 4]
```diff
@@ -228,6 +229,113 @@ def test_inline_hidden_field_no_column(self):
...
+            '<div class="form-row hidden field-position"><div><label>Position:</label>'
+            '<div class="readonly">0</div></div></div>',
+            response.rendered_content)
+        self.assertInHTML(
+            '<div class="form-row hidden field-position"><div><label>Position:</label>'
+            '<div class="readonly">1</div></div></div>',
+            response.rendered_content)
+
+    def test_tabular_inline_with_hidden_field_non_field_errors_has_correct_colspan(self):
+        """#31867 -- in tabuar inlines, when a form has non-field errors, those errors are rendered in a table line
```
Comment (jacobtylerwalls) [inline tests/admin_inlines/tests.py:306]: "tabular", also see above re: boilerplate language, wrapping, and periods.

**Review by jacobtylerwalls — COMMENTED** [review]
Hi @antoinehumbert -- since this is getting closer to a final review, I figured I'd leave some style feedback to make this more mergeable. Hopefully you have some time to polish and address merge conflicts. Regards, Jacob

**Comment by carltongibson** [comment]
I'm reviewing this now. I'd already rebased and resolved the conflicts, so I've pushed those to avoid any duplicate effort.

**Review by carltongibson — COMMENTED** [review]
Hi @antoinehumbert — Right, sorry for the slow pick-up on this one. It's a tough one, so it's taken a while to sit down with it. 

As per the previous discussion, I think improving the consistency is good. I've retitled the PR emphasise that. 

Can I ask now that we cut down the footprint of the PR? There's a lot of duplication in the tests we can get rid of to make the changes clearer. (Adding new models also slows down the test suite, so if we can reuse models that's a good goal.) 

If you look in sibling `admin_views` tests you'll see that we reregister the same models time and again with different `AdminSite` instances in order to avoid creating duplicate models. e.g.: 

https://github.com/django/django/blob/c6b07627fcb5d1c8d2082714ef5adb63bee6cf4c/tests/admin_views/admin.py#L1127-L1134

There's still a lot of models, but we try and reuse where possible. Can we the reuse one of the existing models in `admin_inlines`? (`Parent` and `Child` would do no?) 

Then, we should be able to reduce the `Form` and `*Inline` classes in `admin.py`. (Where we need the second inline with the `fields` declaration we could add a docstring saying why that's needed, so someone, maybe one of us, coming back to this can see immediately what it's for.) 

Finally then, we can move the duplicate `setUp` lines to `setUp` or `setUpTestData`. 

This should leave the actual thing under test much clearer to the eye, and so easier to reason about and maintain. 

I think the overall change is reasonable. Thanks 👍

**Comment by antoinehumbert** [comment]
Hi @carltongibson, I'll have a look at your and @jacobtylerwalls feedbacks tomorrow and make the suggested simplifications in tests.

File: tests/admin_inlines/tests.py [hunk 5]
```diff
@@ -226,6 +227,113 @@ def test_inline_hidden_field_no_column(self):
...
+        )
+        self.assertInHTML(
+            '<div class="fieldBox field-position hidden"><label class="inline">Position:</label>'
+            '<div class="readonly">1</div></div>',
+            response.rendered_content
+        )
+
+    def test_stacked_inline_single_hidden_field_in_line_with_view_only_permissions(self):
+        """#31867 -- Make sure content of hidden field is not visible in stacked inline when user has view-only
+        permission and field is the only field on one line"""
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:280]: Please chop the ticket number. Also, to add more explanation, docstrings should state the expected behavior and omit prefixes like "Make sure" since all tests ensure things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

This comment applies to all new docstrings.

File: tests/admin_inlines/tests.py [hunk 6]
```diff
@@ -226,6 +227,113 @@ def test_inline_hidden_field_no_column(self):
...
+        parent = Parent3.objects.create(name='a')
+        Child3.objects.create(name='b', position='0', parent=parent)
+        Child3.objects.create(name='c', position='1', parent=parent)
+        user = User.objects.create_user(username="user", password="pwd", is_staff=True)
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:284]: Please use single quotes (everywhere), you mixed single with double quotes.

File: tests/admin_inlines/tests.py [hunk 7]
```diff
@@ -226,6 +227,113 @@ def test_inline_hidden_field_no_column(self):
...
+        user.user_permissions.add(permission)
+        permission = Permission.objects.get(codename='view_child3', content_type=child3_ct)
+        user.user_permissions.add(permission)
+        self.client.login(username="user", password="pwd")
+        response = self.client.get(reverse('admin:admin_inlines_parent3_change', args=(parent.pk,)))
+        # Following test ensure that the whole line containing position field is hidden
+        self.assertInHTML(
+            '<div class="form-row hidden field-position"><div><label>Position:</label>'
+            '<div class="readonly">0</div></div></div>',
+            response.rendered_content)
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:297]: Use hanging indentation.

File: tests/admin_inlines/tests.py [hunk 8]
```diff
@@ -226,6 +227,113 @@ def test_inline_hidden_field_no_column(self):
...
+            'child1_set-MIN_NUM_FORMS': 0,
+            'child1_set-MAX_NUM_FORMS': 1000,
+            '_save': 'Save',
+            'child1_set-0-id': child.id,
+            'child1_set-0-parent': parent.id,
+            'child1_set-0-name': child.name,
+            'child1_set-0-position': 1,
+        }
+        response = self.client.post(change_url, data)
+        # Form has 3 visible columns and 1 hidden column
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:324]: Add a trailing period.

File: tests/admin_inlines/admin.py [hunk 9]
```diff
@@ -342,6 +343,59 @@ class ClassAdminStackedVertical(admin.ModelAdmin):
...
+
+
+class Child3Form(forms.ModelForm):
+
+    class Meta:
+        fields = '__all__'
+        model = Child3
+        widgets = {
+            'position': forms.HiddenInput,
+        }
```
Comment (felixxm) [inline tests/admin_inlines/admin.py:390]: This can be single-lined.

File: tests/admin_inlines/admin.py [hunk 10]
```diff
@@ -342,6 +343,59 @@ class ClassAdminStackedVertical(admin.ModelAdmin):
...
+            self.add_error(None, ValidationError("A non-field error"))
+
+
+class Child1Inline(admin.TabularInline):
+    model = Child1
+    form = Child1Form
+
+
+class Child2Form(forms.ModelForm):
+
```
Comment (felixxm) [inline tests/admin_inlines/admin.py:368]: Chop blank line.

File: tests/admin_inlines/admin.py [hunk 11]
```diff
@@ -342,6 +343,59 @@ class ClassAdminStackedVertical(admin.ModelAdmin):
...
+        model = Child2
+        widgets = {
+            'position': forms.HiddenInput,
+        }
+
+
+class Child2Inline(admin.StackedInline):
+    model = Child2
+    form = Child2Form
+    fields = (("name", "position"), )
```
Comment (felixxm) [inline tests/admin_inlines/admin.py:380]: Use list and remove unnecessary whitespace
```suggestion
    fields = [('name', 'position')]
```

File: tests/admin_inlines/models.py [hunk 12]
```diff
@@ -326,6 +326,36 @@ class Meta:
...
+
+
+class Child2(models.Model):
+    name = models.CharField(max_length=1)
+    position = models.PositiveIntegerField()
+    parent = models.ForeignKey(Parent2, models.CASCADE)
+
+
+class Parent3(models.Model):
+    name = models.CharField(max_length=1)
```
Comment (felixxm) [inline tests/admin_inlines/models.py:351]: Do we need so many new models? e.g. it should be enough to have a single "Parent" model and register it with 3 inlines. We could also use better names rather than `Child1/2/3`, e.g. `ChildHiddenFieldsTabular`.

**Review by felixxm — COMMENTED** [review]
@antoinehumbert Thanks :+1: I left stylistic comments.

**Comment by antoinehumbert** [comment]
Hi @carltongibson, @jacobtylerwalls, @felixxm,

I just pushed modifications taking into account your feedbacks.

Let me know if everything is ok

**Review by carltongibson — APPROVED** [review]
Hi @antoinehumbert — thanks for this, it's really good. 

I've rebased, and pushed some small edits in be28cab183fcb1a46b8e3429e68f49574aa5b039: 

* Small adjustment to the tabular.html template, to simplify the conditional logic there. 
* Move the repeated setup logic to `setUpTestData`, which makes the tests easier to read to I think. 
* Small cosmetic changes. 

Let me know what you think. We can squash down and merge. 
(@felixxm may wish to glance this evening? 🙂) 

Good work! ⛵️

File: django/contrib/admin/templatetags/admin_modify.py [hunk 13]
```diff
@@ -106,10 +106,11 @@ def cell_count(inline_admin_form):
     """Return the number of cells used in a tabular inline."""
     count = 1  # Hidden cell with hidden 'id' field
     for fieldset in inline_admin_form:
-        # Loop through all the fields (one per cell)
+        # Loop through all the visible fields (one per cell)
```
Comment (felixxm) [inline django/contrib/admin/templatetags/admin_modify.py:109]:
```suggestion
        # Count all visible fields.
```

File: tests/admin_inlines/admin.py [hunk 14]
```diff
@@ -342,6 +342,36 @@ class ClassAdminStackedVertical(admin.ModelAdmin):
     inlines = [ClassStackedVertical]
 
 
+# Admins and form for #31867.
+class ChildWithHiddenFieldForm(forms.ModelForm):
+    class Meta:
+        fields = ['name', 'position', 'parent']
+        model = SomeChildModel
+        widgets = {'position': forms.HiddenInput}
```
Comment (felixxm) [inline tests/admin_inlines/admin.py:350]:
```suggestion
        model = SomeChildModel
        fields = ['name', 'position', 'parent']
        widgets = {'position': forms.HiddenInput}
```

File: tests/admin_inlines/admin.py [hunk 15]
```diff
@@ -342,6 +342,36 @@ class ClassAdminStackedVertical(admin.ModelAdmin):
...
+
+    def _post_clean(self):
+        super()._post_clean()
+        if self.instance is not None and self.instance.position == 1:
+            self.add_error(None, ValidationError("A non-field error"))
```
Comment (felixxm) [inline tests/admin_inlines/admin.py:355]:
```suggestion
            self.add_error(None, ValidationError('A non-field error'))
```

File: tests/admin_inlines/admin.py [hunk 16]
```diff
@@ -342,6 +342,36 @@ class ClassAdminStackedVertical(admin.ModelAdmin):
     inlines = [ClassStackedVertical]
 
 
+# Admins and form for #31867.
```
Comment (felixxm) [inline tests/admin_inlines/admin.py:345]: I think we can chop this comment.

File: tests/admin_inlines/tests.py [hunk 17]
```diff
@@ -226,6 +238,110 @@ def test_inline_hidden_field_no_column(self):
...
+                'stacked_inline_with_hidden_field_in_group_admin:admin_inlines_someparentmodel_change',
+                args=(self.parent.pk,),
+            )
+        )
+        # The whole line containing name + position fields is not hidden.
+        self.assertContains(response, '<div class="form-row field-name field-position">')
+        # The div containing the position field is hidden.
+        self.assertInHTML(
+            '<div class="fieldBox field-position hidden"><label class="inline">Position:</label>'
+            '<div class="readonly">0</div></div>',
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:274]:
```suggestion
            '<div class="fieldBox field-position hidden">'
            '<label class="inline">Position:</label>'
            '<div class="readonly">0</div></div>',
```

File: tests/admin_inlines/tests.py [hunk 18]
```diff
@@ -226,6 +238,110 @@ def test_inline_hidden_field_no_column(self):
...
+            response.rendered_content
+        )
+        self.assertInHTML(
+            '<div class="fieldBox field-position hidden"><label class="inline">Position:</label>'
+            '<div class="readonly">1</div></div>',
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:279]:
```suggestion
            '<div class="fieldBox field-position hidden">'
            '<label class="inline">Position:</label>'
            '<div class="readonly">1</div></div>',
```

File: tests/admin_inlines/tests.py [hunk 19]
```diff
@@ -226,6 +238,110 @@ def test_inline_hidden_field_no_column(self):
...
+            'somechildmodel_set-0-id': child.id,
+            'somechildmodel_set-0-parent': parent.id,
+            'somechildmodel_set-0-name': child.name,
+            'somechildmodel_set-0-position': 1,
+        }
+        response = self.client.post(change_url, data)
+        # Form has 3 visible columns and 1 hidden column.
+        self.assertInHTML(
+            '<thead><tr><th class="original"></th><th class="column-name required">Name</th>'
+            '<th class="column-position required hidden">Position</th><th>Delete?</th></tr></thead>',
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:335]:
```suggestion
            '<thead><tr><th class="original"></th>'
            '<th class="column-name required">Name</th>'
            '<th class="column-position required hidden">Position</th>'
            '<th>Delete?</th></tr></thead>',
```

File: tests/admin_inlines/tests.py [hunk 20]
```diff
@@ -226,6 +238,110 @@ def test_inline_hidden_field_no_column(self):
...
+            response.rendered_content
+        )
+        # The non-field error must be spanned on 3 (visible) columns.
+        self.assertInHTML(
+            '<tr class="row-form-errors"><td colspan="3"><ul class="errorlist nonfield">'
+            '<li>A non-field error</li></ul></td></tr>',
```
Comment (felixxm) [inline tests/admin_inlines/tests.py:341]:
```suggestion
            '<tr class="row-form-errors"><td colspan="3">'
            '<ul class="errorlist nonfield"><li>A non-field error</li></ul></td></tr>',
```

**Review by felixxm — COMMENTED** [review]
@antoinehumbert Thanks :+1: 

I left tiny suggestions.

**Comment by felixxm** [comment]
@antoinehumbert Welcome aboard :boat:

**Comment by felixxm** [comment]
buildbot, test on selenium.
